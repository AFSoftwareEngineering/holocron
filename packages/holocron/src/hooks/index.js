export {
  default as useHolocron,
  useHolocronContext,
  Holocron,
  HolocronContext,
} from './useHolocron';
