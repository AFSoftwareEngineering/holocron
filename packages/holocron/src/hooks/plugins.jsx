import React, { Profiler, StrictMode, cloneElement } from 'react';

export function ProfilingPlugin({ moduleName, onRender }) {
  return (children) => (
    <Profiler id={moduleName} onRender={onRender}>
      {children}
    </Profiler>
  );
}

export function StrictModePlugin({ moduleName }) {
  return (children) => (
    <StrictMode key={moduleName}>
      {children}
    </StrictMode>
  );
}

export function localStoragePlugin({ children }) {
  return cloneElement(children, {
    localStorage: {
      getItem: () => null,
    },
  });
}
