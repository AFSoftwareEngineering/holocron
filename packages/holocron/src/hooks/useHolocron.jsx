/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

/* eslint-disable no-underscore-dangle, react/jsx-pascal-case */

import React, { useMemo } from 'react';
import { Provider } from 'react-redux';
import PropTypes from 'prop-types';

import useModuleRegistry from './useModuleRegistry';
import createHolocronStore from '../createHolocronStore';

export const HolocronContext = React.createContext({});

export function useHolocronContext() {
  return React.useContext(HolocronContext);
}

export default function useHolocron(options = {}) {
  const {
    // store config
    store: providedStore,
    reducer = (state) => state,
    initialState,
    enhancer,
    localsForBuildInitialState,
    extraThunkArguments,
    // registry config
    blockedModules,
    holocronModules,
    holocronModuleMap,
    // settings
    ssr = false,
  } = options;

  const initialized = React.useRef(false);
  const registry = useModuleRegistry(holocronModuleMap, holocronModules, blockedModules);
  const store = React.useRef(providedStore || createHolocronStore({
    reducer,
    initialState,
    enhancer,
    localsForBuildInitialState,
    extraThunkArguments: { ...extraThunkArguments, registry },
  }));

  React.useEffect(() => {
    if (!initialized.current) initialized.current = true;
    store.current.rebuildReducer(registry);
  }, [registry]);

  const context = useMemo(
    () => ({ store: store.current, registry, ssr }),
    [store, registry, ssr]
  );

  return React.useMemo(() => {
    function __Holocron__({ children }) {
      return (
        <Provider store={store.current}>
          <HolocronContext.Provider value={context}>
            {initialized && children}
          </HolocronContext.Provider>
        </Provider>
      );
    }
    __Holocron__.propTypes = {
      children: PropTypes.node,
    };
    __Holocron__.defaultProps = {
      children: null,
    };
    __Holocron__.getContext = () => context;
    return __Holocron__;
  }, [context, initialized]);
}

export function Holocron({ children, ...props }) {
  const __Holocron__ = useHolocron(props);
  return (
    <__Holocron__>
      {children}
    </__Holocron__>
  );
}
Holocron.propTypes = {
  children: PropTypes.node,
};
Holocron.defaultProps = {
  children: null,
};
