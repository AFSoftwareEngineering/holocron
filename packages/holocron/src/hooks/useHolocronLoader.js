/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import { useState, useMemo, useEffect } from 'react';
import { useDispatch, batch } from 'react-redux';

import { getModule } from '../moduleRegistry';
import { loadModule } from '../ducks/load';

export default function useHolocronLoader(moduleAny, modules) {
  const dispatch = useDispatch();

  let moduleName;

  const [status, setStatus] = useState('initial');
  const [LoadedModule, setLoadedModule] = useState(() => {
    if (Array.isArray(moduleAny)) {
      const [Module] = moduleAny;
      return [Module];
    }
    return typeof moduleAny === 'string' ? getModule(moduleAny, modules) : null;
  });
  const [Module, setModule] = useState(() => LoadedModule || null);
  const config = useMemo(() => (Module ? Module.holocron : {}), [Module]);

  useEffect(() => {
    if (!LoadedModule) {
      setStatus('loading');
      batch(() => {
        dispatch(loadModule(moduleName)).then(setLoadedModule).catch(() => setStatus('error'));
      });
    } else {
      setModule(LoadedModule);
      setStatus('loaded');
    }
  }, [LoadedModule]);

  return [Module, config, status];
}
