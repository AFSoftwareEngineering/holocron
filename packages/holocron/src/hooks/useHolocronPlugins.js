import { cloneElement, useMemo, useState } from 'react';

// Streamline a mass upgrade with Holocron plugins.
// A plugin is React component that returns its children (or shut off a branch)
// We can asynchronously load data and prop it to our holocron module.
// HoCs work real well with plugins, they can be used as decorators.

// security helmet
// analytics add on
// intl add on
// profiling
// error boundaries
// lib integrations - one service worker
// enhancers wrap the base of the root, while plugins are registered by moduleName
export default function useHolocronPlugins(initialPlugins = []) {
  const [plugins, setPlugins] = useState(() => new Set([].concat(initialPlugins)));

  return useMemo(() => ({
    store: plugins,
    getEnhancers: () => [...plugins].filter((plug) => !!plug.enhancer),
    getModulePlugins: (moduleName) => [...plugins].filter((plug) => plug.moduleName === moduleName),
    addPlugin: ({
      moduleName, plugin, enhancer,
    }) => {
      setPlugins(new Set([...plugins, {
        moduleName, plugin, enhancer,
      }]));
    },
    removePlugin: () => null,
    renderPlugins: ({ children, plugins: plugs }) => cloneElement(children, { plugins: plugs }),
    renderModulePlugins: (children) => children,
    renderEnhancers: (children) => children,
  }), [plugins]);
}
