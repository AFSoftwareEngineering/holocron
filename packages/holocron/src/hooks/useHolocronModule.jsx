/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

/* eslint-disable
  camelcase,
  react/jsx-pascal-case,
  no-underscore-dangle,
  react/jsx-props-no-spreading,
  react/destructuring-assignment
*/

import React from 'react';
import PropTypes from 'prop-types';
import hoistNonReactStatics from 'hoist-non-react-statics';
import { useSelector, useDispatch } from 'react-redux';
import shallowEqual from 'shallowequal';

import { MODULE_REDUCER_ADDED } from '../ducks/constants';
import { loadModule, registerModuleReducer } from '../ducks/load';
import { useHolocronContext } from './useHolocron';
import {
  getLoadModuleDataFn,
  getModuleReducer,
  createModuleStateSelector,
  createHolocronModuleStateSelector,
} from '../utility';

export const HolocronModuleContext = React.createContext({});

export function useHolocronModuleContext() {
  return React.useContext(HolocronModuleContext);
}

export default function useHolocronModule(config = { name: 'module' }) {
  const { store, registry, ssr } = useHolocronContext();

  const __Holocron_Module__ = React.useMemo(
    () => registry.getModule(config.name) || null,
    [config, registry]
  );

  const moduleConfig = React.useMemo(() => ({
    ...(__Holocron_Module__ || {}).holocron || {},
    ...config || {},
    reducer: config.reducer || getModuleReducer(__Holocron_Module__),
    loadModuleData: config.loadModuleData || getLoadModuleDataFn(__Holocron_Module__),
  }), [config, registry, __Holocron_Module__]);

  const moduleState = useSelector(
    React.useMemo(
      () => createModuleStateSelector(moduleConfig),
      [moduleConfig]
    )
  );
  const holocronState = useSelector(
    React.useMemo(
      () => createHolocronModuleStateSelector(moduleConfig),
      [moduleConfig]
    )
  );

  const dispatch = useDispatch();

  const [moduleData, setModuleData] = React.useState(null);
  const [moduleLoadStatus, setLoadStatus] = React.useState(() => (ssr ? 'loaded' : 'initial'));
  const loadModuleData = React.useCallback((ownProps) => {
    if (!moduleConfig.loadModuleData || ssr) {
      setLoadStatus('loaded');
      return Promise.resolve();
    }
    setLoadStatus('loading');
    return Promise.resolve(
      dispatch(async (_, getState, { fetchClient }) => moduleConfig
        .loadModuleData({
          store: { dispatch, getState },
          fetchClient,
          ownProps,
          module: __Holocron_Module__,
        }))
    ).then((data) => {
      if (process.env.NODE_ENV === 'development') {
        console.debug('Holocron Module "%s" data loaded successfully!', moduleConfig.name);
        console.dir(data);
      }
      setLoadStatus('loaded');
      setModuleData(data);
      return data;
    }).catch((error) => {
      if (process.env.NODE_ENV === 'development') {
        console.debug('Holocron Module "%s" data failed to be loaded..', moduleConfig.name);
        console.error(error);
      }
      setLoadStatus('error');
      return error;
    });
  }, [moduleConfig, __Holocron_Module__, ssr]);

  React.useEffect(() => {
    if (
      moduleConfig.reducer
      && registry.getModuleReducer(moduleConfig.name) !== moduleConfig.reducer
    ) {
      if (process.env.NODE_ENV === 'development') {
        console.debug('Loading reducer for Holocron Module "%s"', moduleConfig.name);
      }
      // TODO: rebuild reducer for module
      registry.setModuleReducer(moduleConfig.name, moduleConfig.reducer);
    }
  }, [moduleConfig]);

  React.useEffect(() => {
    if (
      !holocronState.isReducerLoaded && registry.getModuleReducer(moduleConfig.name)
    ) {
      if (process.env.NODE_ENV === 'development') {
        console.debug('Loading reducer for Holocron Module "%s"', moduleConfig.name);
      }
      dispatch(registerModuleReducer(moduleConfig.name));
      store.rebuildReducer(registry);
      dispatch({ type: MODULE_REDUCER_ADDED });
    }
  }, [registry]);

  React.useLayoutEffect(() => {
    if (process.env.NODE_ENV === 'development') {
      if (typeof moduleConfig.name !== 'string' || !moduleConfig.name) {
        console.assert(false, 'Expected "Module.holocron.name" to be configured, received "%s"', moduleConfig.name);
      } else {
        console.debug('Holocron Module "%s" started successfully!', moduleConfig.name);
        console.dir(moduleConfig);
      }
    }

    dispatch(loadModule(moduleConfig.name, registry, moduleConfig.moduleMetaData))
      .then((module) => {
        if (process.env.NODE_ENV === 'development') {
          console.debug('Holocron Module "%s" loaded successfully!', moduleConfig.name);
          console.dir(moduleState);
        }
        if (!registry.getModule(moduleConfig.name)) {
          registry.registerModule(moduleConfig.name, module);
        }
      });
  }, [__Holocron_Module__]);

  React.useEffect(() => {
    if (holocronState.isHolocronLoaded) loadModuleData({});
  }, [loadModuleData, holocronState]);

  const context = React.useMemo(() => ({
    holocronState,
    moduleConfig,
    moduleState,
    moduleData,
    moduleLoadStatus,
    loadModuleData,
  }), [
    holocronState,
    moduleConfig,
    moduleState,
    moduleData,
    moduleLoadStatus,
    loadModuleData,
  ]);

  // TODO: integrate layers here - useHolocronLayers(...);
  return React.useMemo(
    () => {
      const Module = __Holocron_Module__ || (() => null);
      const MemoizedModule = hoistNonReactStatics(
        React.memo(
          (props) => (
            <HolocronModuleContext.Provider value={context}>
              {__Holocron_Module__ ? (
                <__Holocron_Module__ {...props} {...context} />
              // eslint-disable-next-line react/prop-types
              ) : props.children || null}
            </HolocronModuleContext.Provider>
          ),
          typeof moduleConfig.shouldModuleReload === 'function'
            ? (prevProps, nextProps) => !moduleConfig.shouldModuleReload(prevProps, nextProps)
            : shallowEqual
        ),
        Module
      );
      MemoizedModule.displayName = `HolocronModule(${Module.displayName || Module.name})`;
      MemoizedModule.holocron = moduleConfig;
      MemoizedModule.getContext = () => context;
      return MemoizedModule;
    },
    [context, __Holocron_Module__]
  );
}

export function HolocronModule({
  name,
  reducer,
  loadModuleData,
  shouldModuleReload,
  initialState,
  moduleMetaData,
  ...props
}) {
  const __Holocron_Module__ = useHolocronModule(
    React.useMemo(() => ({
      name, reducer, loadModuleData, shouldModuleReload, initialState, moduleMetaData,
    }), [name, reducer, loadModuleData, shouldModuleReload, initialState, moduleMetaData])
  );
  return (
    <__Holocron_Module__ {...props} />
  );
}
HolocronModule.propTypes = {
  name: PropTypes.string,
  reducer: PropTypes.func,
  loadModuleData: PropTypes.func,
  shouldModuleReload: PropTypes.func,
  initialState: PropTypes.shape({}),
  moduleMetaData: PropTypes.shape({}),
};
HolocronModule.defaultProps = {
  name: 'module',
  reducer: undefined,
  loadModuleData: undefined,
  shouldModuleReload: undefined,
  initialState: undefined,
  moduleMetaData: undefined,
};
