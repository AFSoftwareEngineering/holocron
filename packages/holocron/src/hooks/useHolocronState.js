/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

import {
  useMemo, useEffect, useState, useCallback,
} from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { createSelector } from 'reselect';

import { HOLOCRON_STORE_KEY } from '../ducks/constants';
import { getInitialState, loadModule } from '../ducks/load';
import { validateModule } from '../utility';
import { useHolocronContext } from './useHolocron';

export default function useHolocronState(moduleName) {
  const holocron = useHolocronContext();
  const holocronStateSelector = useMemo(() => createSelector(
    (state) => state.getIn(
      [HOLOCRON_STORE_KEY],
      getInitialState()
    ),
    (holocronState) => ({
      isModuleRegistered: !!holocron.registry.getModule(moduleName),
      isModuleBlocked: !!holocron.registry.isModuleBlocked(moduleName),
      isReducerLoaded: holocronState.getIn(['withReducers', moduleName], false),
      isHolocronLoaded: holocronState.getIn(['loaded', moduleName], false),
      isHolocronLoading: holocronState.getIn(['loading', moduleName], false),
      isHolocronError: holocronState.getIn(['failed', moduleName], false),
    })
  ), [moduleName, holocron.registry]);
  const holocronModuleState = useSelector(holocronStateSelector);
  const { isModuleRegistered } = holocronModuleState;

  const dispatch = useDispatch();
  const [Module, setModule] = useState(() => holocron.registry.getModule(moduleName));

  const loadHolocronModule = useCallback(() => {
    if (Module) return Module;

    const promise = dispatch(loadModule(moduleName, holocron.registry))
      .then((LoadedModule) => {
        if (LoadedModule) {
          if (!isModuleRegistered) holocron.registry.registerModule(moduleName, LoadedModule);
          setModule(validateModule(LoadedModule));
        }
        return LoadedModule;
      });
    return promise;
  }, [moduleName, holocron.registry]);

  useEffect(() => {
    if (moduleName) loadHolocronModule();
  }, [moduleName]);

  return [Module, holocronModuleState, loadHolocronModule];
}
