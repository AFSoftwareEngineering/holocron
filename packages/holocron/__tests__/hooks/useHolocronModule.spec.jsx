/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

/* eslint-disable react/prop-types, import/no-extraneous-dependencies  */
import React from 'react';
import { fromJS } from 'immutable';

import { useDispatch } from 'react-redux';
import useHolocronModule, { HolocronModule, useHolocronModuleContext } from '../../src/hooks/useHolocronModule';
import { Holocron } from '../../src/hooks/useHolocron';

import { render, act, waitFor } from '../__util__/renderer';

jest.mock('../../src/loadModule.web', () => ({ default: () => Promise.resolve(() => null) }));

beforeEach(() => {
  jest.clearAllMocks();
});

beforeAll(() => {
  jest.spyOn(console, 'assert').mockImplementation();
  jest.spyOn(console, 'dir').mockImplementation();
  // jest.spyOn(console, 'debug').mockImplementation();
});

const TestModuleContext = () => {
  const {
    moduleLoadStatus, moduleState, moduleData, moduleConfig, holocronState, loadModuleData,
  } = useHolocronModuleContext();
  return (
    <p>
      {JSON.stringify({
        moduleLoadStatus, moduleState, moduleData, moduleConfig, holocronState, loadModuleData,
      })}
    </p>
  );
};

const TestModule = ({
  moduleLoadStatus, moduleState, moduleData, moduleConfig, holocronState, loadModuleData,
}) => (
  <p>
    {JSON.stringify({
      moduleLoadStatus, moduleState, moduleData, moduleConfig, holocronState, loadModuleData,
    })}
  </p>
);

describe(useHolocronModule.name, () => {
  const TestUseHolocronModule = ({ children }) => {
    const HolocronModuleContainer = useHolocronModule();
    return (
      <HolocronModuleContainer>
        {children}
      </HolocronModuleContainer>
    );
  };

  test('uses holocron context in sub component to access module state', async () => {
    expect(render((
      <Holocron>
        <TestUseHolocronModule>
          <TestModuleContext />
        </TestUseHolocronModule>
      </Holocron>
    )).asFragment()).toMatchSnapshot();
  });
});

describe(HolocronModule.name, () => {
  const moduleMetaData = { browser: { url: 'https://example.com/modules/module.js' } };

  beforeAll(() => {
    global.BROWSER = true;
  });

  test('attempts to load up module data without "loadModuleData"', async () => {
    const loadModuleData = jest.fn();
    const Module = ({ children, ...props }) => React.cloneElement(children, props);

    let result;
    await act(async () => {
      result = render(
        <Holocron
          holocronModules={{ Module }}
        >
          <HolocronModule
            name="Module"
            moduleMetaData={moduleMetaData}
          >
            <TestModule />
          </HolocronModule>
        </Holocron>
      );
      await waitFor(() => result.findByText(/"moduleLoadStatus":"loaded"/));
    });

    expect(loadModuleData).toHaveBeenCalledTimes(0);
    expect(result.asFragment()).toMatchSnapshot();
  });

  test('fails to load up module data using "loadModuleData"', async () => {
    const loadModuleData = jest.fn();
    const Module = ({ children, ...props }) => React.cloneElement(children, props);

    loadModuleData.mockImplementationOnce(() => Promise.reject());

    let result;
    await act(async () => {
      result = render(
        <Holocron
          holocronModules={{ Module }}
        >
          <HolocronModule
            name="Module"
            loadModuleData={loadModuleData}
            moduleMetaData={moduleMetaData}
          >
            <TestModule />
          </HolocronModule>
        </Holocron>
      );
      await waitFor(() => result.findByText(/"moduleLoadStatus":"error"/));
    });

    expect(loadModuleData).toHaveBeenCalledTimes(1);
    expect(result.asFragment()).toMatchSnapshot();
  });

  test('loads up module data with "loadModuleData"', async () => {
    const loadModuleData = jest.fn();
    const Module = ({ children, ...props }) => React.cloneElement(children, props);

    let result;
    await act(async () => {
      result = render(
        <Holocron
          holocronModules={{ Module }}
        >
          <HolocronModule
            name="Module"
            moduleMetaData={moduleMetaData}
            loadModuleData={loadModuleData}
          >
            <TestModule />
          </HolocronModule>
        </Holocron>
      );
      await waitFor(() => result.findByText(/"moduleLoadStatus":"loaded"/));
    });

    expect(loadModuleData).toHaveBeenCalledTimes(1);
    expect(result.asFragment()).toMatchSnapshot();
  });

  test('using a reducer supplied to a module', async () => {
    const reducer = jest.fn((state = fromJS({ count: 0 }), action) => {
      if (action.type === 'increment') return state.update('count', (current) => current + 1);
      return state;
    });
    reducer.buildInitialState = () => fromJS({ count: 1 });
    const increment = () => ({ type: 'increment' });
    const TestReduxReducer = () => {
      const dispatch = useDispatch();
      React.useEffect(() => {
        dispatch(increment());
      }, []);
      return null;
    };
    let result;
    await act(async () => {
      result = render(
        <Holocron>
          <HolocronModule
            reducer={reducer}
            initialState={fromJS({ count: 1 })}
          >
            <TestModule />
            <TestModuleContext />
            {/* <TestReduxReducer /> */}
          </HolocronModule>
        </Holocron>
      );
    });

    await act(async () => {
      await waitFor(() => result.findByText(/"isReducerLoaded":"module"/));
    });

    expect(result.asFragment()).toMatchSnapshot();
  });
});
