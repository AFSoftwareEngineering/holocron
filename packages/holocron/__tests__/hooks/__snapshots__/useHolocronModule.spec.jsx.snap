// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HolocronModule attempts to load up module data without "loadModuleData" 1`] = `
<DocumentFragment>
  <p>
    {"moduleLoadStatus":"loaded","moduleState":null,"moduleData":null,"moduleConfig":{"name":"Module","moduleMetaData":{"browser":{"url":"https://example.com/modules/module.js"}}},"holocronState":{"isReducerLoaded":false,"isHolocronLoaded":"Module","isHolocronLoading":false,"isHolocronError":false}}
  </p>
</DocumentFragment>
`;

exports[`HolocronModule fails to load up module data using "loadModuleData" 1`] = `
<DocumentFragment>
  <p>
    {"moduleLoadStatus":"error","moduleState":null,"moduleData":null,"moduleConfig":{"name":"Module","moduleMetaData":{"browser":{"url":"https://example.com/modules/module.js"}}},"holocronState":{"isReducerLoaded":false,"isHolocronLoaded":"Module","isHolocronLoading":false,"isHolocronError":false}}
  </p>
</DocumentFragment>
`;

exports[`HolocronModule loads up module data with "loadModuleData" 1`] = `
<DocumentFragment>
  <p>
    {"moduleLoadStatus":"loaded","moduleState":null,"moduleConfig":{"name":"Module","moduleMetaData":{"browser":{"url":"https://example.com/modules/module.js"}}},"holocronState":{"isReducerLoaded":false,"isHolocronLoaded":"Module","isHolocronLoading":false,"isHolocronError":false}}
  </p>
</DocumentFragment>
`;

exports[`HolocronModule using a reducer supplied to a module 1`] = `
<DocumentFragment>
  <p>
    {}
  </p>
  <p>
    {"moduleLoadStatus":"loaded","moduleState":{"count":1},"moduleData":null,"moduleConfig":{"name":"module","initialState":{"count":1}},"holocronState":{"isReducerLoaded":"module","isHolocronLoaded":"module","isHolocronLoading":false,"isHolocronError":false}}
  </p>
</DocumentFragment>
`;

exports[`useHolocronModule uses holocron context in sub component to access module state 1`] = `
<DocumentFragment>
  <p>
    {"moduleLoadStatus":"initial","moduleState":null,"moduleData":null,"moduleConfig":{"name":"module"},"holocronState":{"isReducerLoaded":false,"isHolocronLoaded":false,"isHolocronLoading":{},"isHolocronError":false}}
  </p>
</DocumentFragment>
`;
